Added Status Menu to UI

HeriLFIU opened the pull request:
Closes kytos-ng/topology#125

### Summary

The status menu UI was added. It allows the user to see the status of switches, links, and interfaces. They can filter by name and different parameters such as status, reason, and enabled. 

They can also store table elements in a separate list. When clicking on the Add/Remove button, the element will be added to the approved list. You can view all elements added to the approved list by clicking on the approved button. This allows users to store and organize data within the UI itself.

Data from this table is also saved within local storage, so you can refresh the page and keep all of your filters/data saved.

Users will also be able to click on switches within the table to access their corresponding data.

<img width="692" alt="image" src="https://github.com/kytos-ng/ui/assets/142065709/f13c7404-1761-46d3-a382-957cac50830b">

This new menu can be accessed by pressing `ctrl + shift + s`.

### Local Tests

A combination of all parameters were used all at once to ensure functionality.

To test the new local storage functionality the page was refreshed and the data stayed the same.

### Quality of Life Changes

QoL changes for the mef-eline EVC table will be available through a separate PR.

HeriLFIU commented:
Just fixed two issues I found this morning:

1. Since the `approved list` section was based on the entire `object` this meant that whenever one section changed (such as `active`), it would count as a completely different `object` and add it to the `approved list`, even though it referred to the same `switch, interface, or link`. To solve this I made the `approved list` section based on `id`, so you should no longer be able to add the exact same element twice.
2. The status reason section was just a list, and it looked a bit out of place. Because of this I turned it into a string so that it doesn't look out of place.

viniarck reviewed (commented):
@HeriLFIU, first of all, nicely done how you managed to refine the requirements, and come up with a great implementation. Functionality -wise, overall looks good, but there are some points that need to be reassessed in point of view (or let me know if you've discussed them with the other folks):

1) Whenever this component gets rendered, it doesn't refresh its content, so, it can still display stalled or old data. For instance, here it's still showing links as DOWN even if they are UP later on, maybe whenever a ctrl+shift+s is hit again it refreshes the data while still loading the filters or a add new button to refresh it, probably the former is cleaner? Let me know what you think or if you have suggestions:

![20240422_161700](https://github.com/kytos-ng/ui/assets/1010796/442fb99e-6b05-4050-97c3-305ef1fb970c)

2) Great use of localStorage, should the keys get namespaced with `kytos/ui` though? and then if any other NApp needs to use localStorage they also use `napp/username`, and since this in the core we always use `kytos/ui`? To avoid potential conflicts in the future and overall organization?

3) The changelog hasn't been updated
4) I found the approved feature, interesting but a bit confusing, meaning that it's acting like a "pre-select" filter, but it's not composing with other filters. If the idea was to have a sort of a pre-selection to only show them, the others filters should also compose with it? I found a bit unnatural when I "approved" but then clicked on other filters and then it displyed other non approved items. Maybe I'm misunderstanding the feature, if you could clarify or think if it makes sense.

5) The links name filter apparently isn't matching correctly, shouldn't this "200" string match below?

![20240422_162749](https://github.com/kytos-ng/ui/assets/1010796/b3072661-514b-4d3e-940d-013e3d595bd8)
![20240422_162757](https://github.com/kytos-ng/ui/assets/1010796/ccf68259-ed42-4a7c-8ca3-1e67bda38179)

HeriLFIU commented:
1. Yes, that is a good idea, ill add both a refresh button and also refresh it on ctrl+shift+s, because when someone wants to access it again the data should be updated.
2. I do think so too, since the localStorage may be useful for other UI elements as well.
3. I forgot about that, ill get to it now
4. Yes, its like "pre selected" elements that you store in a separate list. I think that I should create a new section for the normal table and one with these "pre selected" elements, so that it doesn't interfere with the normal filters, but yes you are correct on its function.
5. Ill need to fix the search function, I got mef elines search function to work, but I tried to do something different and so I switched it out. I'll just use mef eline's search function again.
@viniarck

rmotitsuki commented:
Great PR.
If I can suggest something, it would be to add comments to each function, especially those that have Regex.

HeriLFIU commented:
@viniarck I made all of the changes you had mentioned. I also tested it all out and it seems to be working fine without any issue.

@rmotitsuki Ill try and follow some examples and add more comments. I also saw some unique javascript comments being used in some other code, ill try and utilize them as well.

viniarck reviewed (commented):
@HeriLFIU much appreciated the fixes. It's pretty much ready to land. Very powerful and convenient interface. When using the text search I ended up with some sub strings not matching, check out the thread that I opened, and see if it was supposed to be that way or if we can have a contains string match?

HeriLFIU commented:
@rmotitsuki I added the comments.

@viniarck I tested it out and the search function seems to be working properly now.

viniarck reviewed (approved):
Fantastic contribution @HeriLFIU, very neat and sharp UI, great documentation too. I'll leave it approved, but let's wait for network engineers to give their feedback. Italo said he'll review when he can, and if there's nothing else to add/change it'll get shipped. 

![img](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExdGUyY2ZrbnFtZWg5aTZ6ZDZ6eHpjOHBmdzhsYTIxdmQ5YmxtM2VmbSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/26xBEamXwaMSUbV72/giphy.gif)

italovalcy commented:
Hi @HeriLFIU excellent work you have done implementing this feature! It is indeed very useful for our operation!! @viniarck thanks/congrats for the support reviewing and contributing with this feature as well.

Trying to contribute with you guys have done, I have a few commends/suggestions below:
- if possible, can you please add the same filters similar to the Name column into the Status, Enabled and Active columns. That would allow us to filter for a certain name (substring) plus enabled plus active=down
   - same behavior can be accomplished if the filters button allow multiple selection. Although I think having in the table we already got used to (from mef_eline List circuits UI)
- Can you please add the Status menu on the vertical left bar? I know you can just run CTRL+SHIFT+s to see it, but we will easily forget this shortcut and having it on the side bar helps in that sense
- I observed on the interface table you have a second level of scroll bar, Can we remove that and leave only one scrolling to the whole page?
- I tested with the Interface.metadata port_name and it is not showing on the table. Can we also have this similar to what we have for Link.metadata.link_name and Switch.metadata.node_name ?
- The little eye on the "Hidden/Selected" column was not immediate for me the idea. Maybe we could change a little bit and move that column to be the first one (no column title) and change to use a single checkbox (instead of the eye). Then we move the Data>Original/Pre-selected to the filters and renamed them to "Show-Selected" and "Show-All". What do you think?

HeriLFIU commented:
Hello @italovalcy,

1. That makes sense, I can just remove the buttons and replace them with the search columns from the EVC table.
2. I haven't seen the code for the vertical left bar, but I can take a look at it and see how the Status Menu can be incorporated into it.
3. I currently don't know what's causing the scroll bar to appear, maybe it occurs when the table gets too large, but I'll take a look at it and see if I can remove it.
4. I'll add the `port_name` to the menu.
5. The eyes are a very confusing feature, so what you just mentioned is a good way to make it clearer. I'll message you once I finish the changes so that you can see them for yourself and make sure everything is in order.

HeriLFIU commented:
@italovalcy @viniarck 
It took me a while, but everything seems to be working fine now.

I don't know if this should be documented but I added a new way to include UI components within the Menu Bar via events.

To explain it in detail:
1. You create some data that will be sent via an event to the Menu Bar, as seen below.
<img width="490" alt="image" src="https://github.com/kytos-ng/ui/assets/142065709/99d863f8-127e-4535-a0f8-69bd892ab859">

2. You then emit this info via the `includeMenuBar` event, as seen below.
<img width="419" alt="image" src="https://github.com/kytos-ng/ui/assets/142065709/2c027779-3cc5-4c33-b4ca-91f18362393d">

3. Lastly, whenever you click on the corresponding button within the Menu Bar, it will emit another event, the one you had set as `nameCall` within your data. You can then add a listener to receive it and do something with it, such as display the new component.

HeriLFIU commented:
I think this abstracts away a lot of complexity and whenever someone wants to add a new option to the Menu Bar, they can just emit the info from within their component an register a listener to receive something back.

HeriLFIU commented:
I changed `nameCall` to `eventName` since it made more sense

viniarck commented on src/components/kytos/misc/Menubar.vue:
```diff
@@ -14,6 +14,11 @@
                :icon="component.icon"
                :tooltip="component.tooltip"
                @click.native="setItem(index+1)"/>
+     <k-button v-for="(component, index) in separateComponents"
```
@HeriLFIU, I wonder if we could still use the existing `components` instead of adding this new `separateComponents`. I think we can make `KytosStatusMenu` as an official part here too, wdyt? Let me know if you had any other reason what I'm not following here. Also in the `components` it would render before the dynamic NApps ui, so it'd be great too to have it right after the `mapbox-settings`

HeriLFIU replied:
It was giving me a lot of issues, but I just realized its because I forgot to remove the component itself from `APP.vue`, so it would never go away. I just added it to the toolbar and it seems to be working fine now.

HeriLFIU replied:
Now for the key bindings, since its part of the menu bar now, I can just probably set active item to 2.

HeriLFIU replied:
@viniarck actually since its part of the menu bar now, should we have some keybind system where:

The first element could be ctrl+shift+1 then ctrl+shift+2 and so on.

HeriLFIU replied:
because if the menu bar changes, the hot key that was set for active item 2 might also change

viniarck replied:
Interesting, if it's behaving as expected now and having it on `ctrl+shift+2` would be cool. 

Great solution.

viniarck reviewed (commented):
@HeriLFIU, great improvements, most look good to me, It'll leave @italovalcy to double confirm the rest with you. 

Other than that:

1 - Can we have `KytosStatusMenu` on `components`?
2 - Minor issue the selected filter: 

- Select one switch
- Try to filter for a `enabled` value, type something like `"true"`
- Now erase the string on the `enabled` filter it won't show the selected switch again

![20240508_140722](https://github.com/kytos-ng/ui/assets/1010796/279703f7-3b5a-442f-801f-4c1bab0f6944)

![20240508_140732](https://github.com/kytos-ng/ui/assets/1010796/b3fb7f79-6eae-45f9-8c75-5e2867fa156e)

HeriLFIU commented:
@viniarck I resolved the issue with the selected filter, since I wasn't using `structuredClone` for the selected list elements, it was treating it as the same object while filtering and deleting everything within it.

- I fixed the issue with the components and removed what I had added for separate components.

- It was a bit difficult due to the constraints of the v-hotkey package, but I was able to dynamically generate hotkeys based on the number of components within the nav-bar.

- I also cleaned up the code and removed some now unnecessary things, such as the close and maximize button, since the UI will be kept in the Menu Bar.

HeriLFIU commented:
It is ordered from top to bottom within the Menu Bar, so the Users should be able to press `ctrl+shift+"componentNumber"`, to quickly access that element.

For example:

- If the wanna access the map settings, it'll be the first element within the Menu Bar, so it's `ctrl+shift+1`.

- If the wanna access the map Status Menu, it'll be the second element within the Menu Bar, so it's `ctrl+shift+2`.

viniarck commented:
> @viniarck I resolved the issue with the selected filter. Since I wasn't using `structuredClone` for the selected list elements, it was treating it as the same object while filtering and deleting everything within it.
> 
> * I fixed the issue with the components and removed what I had added for separate components.
> * It was a bit difficult due to the constraints of the v-hotkey package, but I was able to dynamically generate hotkeys based on the number of components within the nav-bar.
> * I also cleaned up the code and removed some now unnecessary things, such as the close and maximize button, since the Status Menu  will be kept in the Menu Bar.

Nice @HeriLFIU. Great process and solution. I noticed another unexpected side effect though:

- It looks like the `k-status-menu` is always showing up maximized, where we used to have the initial `mapbox-settings` focused and the topology was displayed. Can you double check if this is also happening for you locally? 

- If you can also update the changelog with `ctrl+shift+2` instead of `ctrl+shift+s`.

HeriLFIU commented:
@viniarck 
It should be working now, I forgot to add the APP.vue within the first commit that added the Status Menu to the Menu Bar.
I also updated the changelog.

viniarck reviewed (commented):
@HeriLFIU appreciated the latest fixes and enhancements, overall it's looking really good. One remaining technical implementation point though:

- Should we have a scrollbar in the tables? For instance here in this case where a topology has 10+ interfaces, it's not showing, unless user starts to filter for the expected string value: 
![20240510_094318](https://github.com/kytos-ng/ui/assets/1010796/13052953-308e-4577-8ad9-42f00215c3f8)

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                       async: true});
+      request.done(function(data) {
+        self.switches = data['topology']['switches']
+        self.links = data['topology']['links']
```
Heads-up: the interface_id, did and content_switch properties on the interface were added because they will be needed for my other comment on the click Event to open the InfoPanel to view/edit the interface details.

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                    </tr>
+                  </thead> 
+                  <tbody>
+                    <tr v-for="(item, index) in filtered_linkData" :key="item.id">
```
Please refer to my comment below regarding opening the InfoPanel to view/edit Link details.

```suggestion
                    <tr v-for="(item, index) in filtered_linkData" :key="item.id" @click="show_infoPanelLink(item.id)">
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                    </tr>
+                  </thead> 
+                  <tbody>
+                    <tr v-for="(item, index) in filtered_interfaceData" :key="item.id">
```
Please refer to my comment below regarding opening the InfoPanel to view/edit Interface details.

```suggestion
                    <tr v-for="(item, index) in filtered_interfaceData" :key="item.id" @click="show_infoPanelInterface(item.id)">
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+  methods: {
+    refresh() {
+      // Refreshes all table data
+      this.get_topology()
+      this.get_interfaces()
```
I see you have two API calls to the topology to get the desired information. I wonder what you think about us trying to optimize this and leave the interface processing to be done at the client side. Please see my comment on the get_topology() method definition.

I'm sorry if you guys already discussed this and decided for going this direction.

```suggestion
      this.get_topology()
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                     title: "Switch Details",
+                     subtitle: this.switches[s].connection}
+      this.$kytos.$emit("showInfoPanel", content)
+    },
```
Adding as a suggestion here some helper function to view/edit Link and Interface details, as you proposed for Switch. I think having those InfoPanel can be very helpful to help the network operator to deploy some changes while still on the Status Menu. Please let me know what you think.

```suggestion
    },
    show_infoPanelLink(l) {
      /**
       * @event showInfoPanel
       * @param {Object} l - The link data from the current row used to generate the info panel.
       * @description Displays info panel when table row is clicked.
       */
      var link = this.links[l]
      var subtitle = ""
      if(link.metadata.link_name !== undefined && link.metadata.link_name.length !== 0){
        subtitle = link.metadata.link_name
      }
      else{
        subtitle = link.id
        if(subtitle.length > 16) subtitle=subtitle.substr(0,16) + "..."
      }
      var content = {component: 'kytos-topology-k-info-panel-link_info',
                     content: link,
                     icon: "cog",
                     title: "Link Details",
                     subtitle: subtitle}
      this.$kytos.$emit("showInfoPanel", content)
    },
    show_infoPanelInterface(i) {
      /**
       * @event showInfoPanel
       * @param {Object} i - The interface data from the current row used to generate the info panel.
       * @description Displays info panel when table row is clicked.
       */
      var content = {component: 'k-interface-info',
                     content: this.interfaces[i],
                     icon: "cog",
                     title: "Interface Details",
                     subtitle: i}
      this.$kytos.$emit("showInfoPanel", content)
    },
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+    }
+  },
+  created() {
+    this.get_topology();
+    this.get_interfaces();
```
As in the above comments, this won't be necessary

```suggestion
    this.get_topology();
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+
+.data_status_table 
+  color: #ccc
+  max-height: 250px
```
Since you removed the overflow-y and overflow-x, this max height was left behind, which was wrapping the table data to only 5 elements (that is why I was not seeing other switch's interfaces ;-) )

```suggestion
```

italovalcy commented on src/components/kytos/misc/Menubar.vue:
```diff
@@ -13,7 +14,7 @@
                :icon="component.icon"
                :tooltip="component.tooltip"
-               @click.native="setItem(index+1)"/>
+               @click.native="setItem(index+1)"/>           
```
```suggestion
               @click.native="setItem(index+1)"/>
```

Removing extra space.

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                  <tbody>
+                    <tr v-for="(item, index) in filtered_interfaceData" :key="item.id">
+                      <td><k-button class="statusTableButton" :icon="isStoredI(item, index)" v-on:click.native="add_remove_Interfaces(item, index)"></k-button></td> 
+                      <td v-on:click="show_infoPanel(item.switch)">{{item.switch}}</td>
```
With the suggestion above, I think we should leave the whole row linked to view/edit Interface information (once viewing/editing the switch is already available above).

```suggestion
                      <td>{{item.switch}}</td>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+        <k-accordion-item title="Refresh">
+          <k-button icon="redo" title="Refresh" tooltip="Refresh" v-on:click.native="refresh()"></k-button>
+        </k-accordion-item>
+        <!-- 
```
```suggestion
        <!--
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                        <input v-model="switchTextFilter[4][1]"></input>
+                      </th>
+                    </tr>
+                  </thead> 
```
```suggestion
                  </thead>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                  </thead> 
+                  <tbody>
+                    <tr v-for="(item, index) in filtered_switchData" :key="item.dpid">
+                      <td><k-button class="statusTableButton" :icon="isStoredS(item, index)" v-on:click.native="add_remove_Switches(item, index)"></k-button></td>  
```
```suggestion
                      <td><k-button class="statusTableButton" :icon="isStoredS(item, index)" v-on:click.native="add_remove_Switches(item, index)"></k-button></td>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                      <td :class="statusColor(item)">{{item.status}}</td>
+                      <td>{{item.status_reason | splitStatusReasons}}</td>
+                      <td>{{item.enabled}}</td>
+                      <td>{{item.active}}</td>              
```
```suggestion
                      <td>{{item.active}}</td>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                      <td>{{item.enabled}}</td>
+                      <td>{{item.active}}</td>              
+                    </tr>
+                  </tbody>             
```
```suggestion
                  </tbody>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                  </thead> 
+                  <tbody>
+                    <tr v-for="(item, index) in filtered_interfaceData" :key="item.id">
+                      <td><k-button class="statusTableButton" :icon="isStoredI(item, index)" v-on:click.native="add_remove_Interfaces(item, index)"></k-button></td> 
```
```suggestion
                      <td><k-button class="statusTableButton" :icon="isStoredI(item, index)" v-on:click.native="add_remove_Interfaces(item, index)"></k-button></td>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                      <td :class="statusColor(item)">{{item.status}}</td>
+                      <td>{{item.status_reason | splitStatusReasons}}</td>
+                      <td>{{item.enabled}}</td>
+                      <td>{{item.active}}</td>           
```
```suggestion
                      <td>{{item.active}}</td>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+                      <td>{{item.enabled}}</td>
+                      <td>{{item.active}}</td>           
+                    </tr>
+                  </tbody>           
```
```suggestion
                  </tbody>
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+    menuBarInfo: {eventName: 'StatusMenu', icon: 'signal', tooltip: 'Status Menu'},
+    currentSwitchData: "",
+    currentLinkData: "",
+    currentInterfaceData: "",     
```
```suggestion
    currentInterfaceData: "",
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+     *              All table data follows the same structure as seen above.
+     *              Split for each individual table: Switches, Links, and Interfaces.
+     */
+    //Base Data 
```
```suggestion
    //Base Data
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+  overflow-y: auto
+  height: 0px
+
+.data_status_table 
```
```suggestion
.data_status_table
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+  padding: 0.45em 0 0.45em 0
+  word-break: break-all
+
+.data_status_table tbody tr:hover 
```
```suggestion
.data_status_table tbody tr:hover
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+  background-color: #666
+
+.yellow
+  background-color: #f9a825 
```
```suggestion
  background-color: #f9a825
```

italovalcy commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,764 @@
+  background-color: #f9a825 
+  color: #222
+
+.green 
```
```suggestion
.green
```

HeriLFIU commented:
@italovalcy @viniarck 
I just finished reading through and committing all of the changes. I also tested them out and found one small error while showing the info panel. From what I understand its trying to access some metadata called `tag_ranges` from the interfaces but it's undefined.
<img width="393" alt="image" src="https://github.com/kytos-ng/ui/assets/142065709/ba61339b-8393-4af2-b675-6301a4ce2569">

viniarck commented on src/components/kytos/misc/StatusMenu.vue:
```diff
@@ -0,0 +1,806 @@
+        self.$kytos.$emit("setNotification", notification)
+      });
+    },
+  //   get_interfaces() {
```
Can you remove this commented method? 

Looks like you meant to remove it, since you're fetching the entire topology.

viniarck reviewed (commented):
LGTM. It'll get merged when Italo's also approve it. 

@HeriLFIU, 

1 - Can you also remove the commented method?
2 - Can you `git squash` your commits into 2? One commit for all the code changes and another for the changelog.rst. At the moment you have 50+ commits, to cherry-pick them for backporting would be too difficult, so you git squash them that'd be great to cherry pick only a few later. Before trying to squash your commits or reseting your branch make sure to create a copy to avoid losing work. Instead of squash you can also git reset to the first commits, and create the two commits and force push. If you need help with git squash or reset let me know.

HeriLFIU commented:
@viniarck 
Thank the lord,
I was trying to find the simplest way to squish everything down into 2 commits, and I read about a method called git interactive rebase.
Luckily, since I had all of my commits to the change log separate from everything else. I was just able to do an interactive rebase, push all of the changelog commits to the back, and then squash all of the other commits up at the front into one.

HeriLFIU commented:
I also removed the commented method.

viniarck reviewed (approved):
Excellent, @HeriLFIU. I'm glad you've managed. These new squashed commits will be trivial to cherry pick and backport.

italovalcy reviewed (approved):
LGTM! Excellent work, @HeriLFIU ! Congratulations!!

viniarck commented:
Let's ship it. 

![img](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcW5kbmVoZWR2eXdnM2tncjRxZDFlaHI5bmE2eWxvZHk0c3dsZjVlbyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/Hw8vYF4DNRCKY/giphy.gif)
